File: javascript/557-reverse-words-in-a-string-iii.js
// Given a string, you need to reverse the order of characters in each word within a sentence while 
// still preserving whitespace and initial word order.

/**
 * @param {string} s
 * @return {string}
 */

// Solution using JavaScript built-in methods to create new array of characters and reverse them.
// Time complexity: O(n), since traversing through the whole array is the dominant term.
// Space complexity: O(n) since you are using a new array of all n characters.
var reverseWords = function(s) {
  return s.split(' ').map(word => {
      return word.split('').reverse().join('');
  }).join(' ');
};

// More compact version of the above solution, but possibly less readible.
var reverseWordsOneLiner = function(s) {
    return s.split(' ').map(word => word.split('').reverse().join('')).join(' ');
};

File: python/557-reverse-words-in-a-string-iii.py
# Given a string, you need to reverse the order of characters in each word within a sentence while 
# still preserving whitespace and initial word order.

# Solution using Python string slicing, map, string split and join builtin methods.
# Time complexity: O(n), since entire string must be iterated through in order to split into array.
# Space complexity: O(n), since map returns an array of n reversed strings.
class Solution:
    def reverseWords(self, s: str) -> str:
        wordArray = s.split(' ')
        reversedWordArray = map(lambda x: x[::-1], wordArray)
        return ' '.join(list(reversedWordArray))